src/devices/Charlieplex/Charlieplex.cs

```diff
@@ -0,0 +1,175 @@
+﻿using System;
+using System.Device.Gpio;
+using System.Threading;
+
+namespace Iot.Device.Multiplex
+{
+    /// <summary>
+    /// Provides support for Charlieplex multiplexing.
+    /// https://wikipedia.org/wiki/Charlieplexing
+    /// </summary>
+    public class Charlieplex : IDisposable
+    {
+        private readonly bool _shouldDispose;
+        private GpioController _controller;
+        private int[] _pins;
+        private int _nodeCount;
+        private CharlieplexNode[] _charlieNodes;
+
+        /// <summary>
+        /// Initializes a new Charlieplex type that can be use for multiplex over a relatively small number of GPIO pins.
+        /// </summary>
+        /// <param name="pins">The set of pins to use.</param>
+        /// <param name="nodeCount">The count of nodes (like LEDs) that will be addressable. If 0, then the Charlieplex maximum is used for the pins provided (n^2-n).</param>
+        /// <param name="gpioController">The GPIO Controller used for interrupt handling.</param>
+        /// <param name="shouldDispose">True (the default) if the GPIO controller shall be disposed when disposing this instance.</param>
+        public Charlieplex(int[] pins, int nodeCount = 0,  GpioController gpioController = null, bool shouldDispose = true)
+        {
+            if (pins.Length < 2)
+            {
+                throw new ArgumentException($"{nameof(Charlieplex)}: 2 or more pins must be provided.");
+            }
+
+            var charlieCount = (int)Math.Pow(pins.Length, 2) - pins.Length;
+            if (nodeCount > charlieCount)
+            {
+                throw new ArgumentException($"{nameof(Charlieplex)}: maximum count is {charlieCount} based on {pins.Length} pins. {nodeCount} was specified as the count.");
+            }
+
+            if (nodeCount == 0)
+            {
+                nodeCount = charlieCount;
+            }
+
+            if (gpioController == null)
+            {
+                gpioController = new GpioController();
+            }
+
+            foreach (var pin in pins)
+            {
+                gpioController.OpenPin(pin, PinMode.Input);
+            }
+
+            _controller = gpioController;
+            _shouldDispose = shouldDispose;
+            _pins = pins;
+            _nodeCount = nodeCount;
+            _charlieNodes = GetCharlieLoads(pins, nodeCount);
+        }
+
+        /// <summary>
+        /// The number of load devices (like LEDs) that can be addressed.
+        /// </summary>
+        public int LoadCount => _nodeCount;
+
+        /// <summary>
+        /// Write a PinValue to a load.
+        /// Address scheme is 0-based. Given 8 loads, addresses would be 0-7.
+        /// </summary>
+        public void Write(int node, PinValue value, int delay = 10)
```

Comment:
Document arguments. And prefer TimeSpan for delays.

Reply:
I tried taking a TimeSpan. It made calling the API so much more verbose. Currently, the API can be called with three ints. that's really nice.

Reply:
I adopted TimeSpan for DisplaySegment. That creates asymmetry, but better accounts for their different use cases.